piet-metal-derive/src/lib.rs

```diff
@@ -94,6 +101,500 @@ impl GpuScalar {
     }
 }
 
+impl std::fmt::Display for GpuScalar {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            GpuScalar::F32 => write!(f, "F32"),
+            GpuScalar::I8 => write!(f, "I8"),
+            GpuScalar::I16 => write!(f, "I16"),
+            GpuScalar::I32 => write!(f, "I32"),
+            GpuScalar::U8 => write!(f, "U8"),
+            GpuScalar::U16 => write!(f, "U16"),
+            GpuScalar::U32 => write!(f, "U32"),
+        }
+    }
+}
+
+// HLSL helpers
+fn hlsl_get_ref_plus_offset(current_offset: usize) -> String {
+    match current_offset {
+        0 => String::from("ref"),
+        _ => format!("ref + {}", current_offset),
+    }
+}
+
+fn hlsl_calculate_num_uints_required_for_storage_of_vector(
```

Comment:
This name is a little over the top. I prefer a reasonable name (storage_size_in_uints maybe?) plus a comment. Also, why a standalone function rather than a method on `GpuScalar`?